ce_expansion/atomgraph/bcm.py

```diff
@@ -0,0 +1,207 @@
+        else:
+            # ensure metal_types to unique, sorted list of metals
+            self.metal_types = sorted(set(metal_types))
+
+        self.metal_pairs = list(itertools.combinations_with_replacement(
+                                    self.metal_types, 2))
+
+        # creating gamma list for every possible atom pairing
+        self.gamma = None
+        self.ce_bulk = None
+        self._get_bcm_params()
+
+        # Calculate and set the precomps matrix
+        self.precomps = None
+        self.precomps = self._calc_precomps()
+
+    def __len__(self):
+        return len(self.atoms)
+
+    def _get_bcm_params(self):
+        """
+            Creates gamma and ce_bulk dictionaries which are then used to created precomputed values for the BCM calculation
+
+        Sets:   Gamma: Weighting factors of the computed elements within the BCM
+                ce_bulk: Bulk Cohesive energy values
+
+        """
+        gamma = {}
+        ce_bulk = {}
+        for item in self.metal_pairs:
+            # Casting metals and setting keys for dictionary
+            metal_1, metal_2 = item
+
+            gamma_obj = GammaValues(metal_1, metal_2)
+
+            # using Update function to create clean Gamma an bulk dictionaries
+            gamma = recursive_update(gamma, gamma_obj.gamma)
+            # add ce_bulk vals
+            ce_bulk[gamma_obj.element_a] = gamma_obj.ce_a
+            ce_bulk[gamma_obj.element_b] = gamma_obj.ce_b
+
+        self.ce_bulk = ce_bulk
+        self.gammas = gamma
+
+    def _calc_precomps(self):
+        """
+            Uses the Gamma and ce_bulk dictionaries to create a precomputed BCM matrix of gammas and ce_bulk values
+
+            [precomps] = [gamma of element 1] * [ce_bulk of element 1 to element 2]
+
+        Returns: Precomp Matrix
+
+        """
+        # precompute values for BCM calc
+        n_met = len(self.metal_types)
+
+        precomps = np.ones((n_met, n_met))
+
+        for i in range(n_met):
+            for j in range(n_met):
+
+                M1 = self.metal_types[i]
+                M2 = self.metal_types[j]
+                precomp_bulk = self.ce_bulk[M1]
+                precomp_gamma = self.gammas[M1][M2]
+
+                precomps[i, j] = precomp_gamma * precomp_bulk
+        return precomps
+
+    def calc_ce(self, orderings):
+        """
+        Calculates the Cohesive energy of the ordering given or of the default ordering of the NP
+
+        [Cohesive Energy] = ( [precomp values of element A and B] / sqrt(12 * CN) ) / [num atoms]
+
+        Args:
+            orderings: The ordering of atoms within the NP; ordering key is based on Metals in alphabetical order
+                - Will use ardering defined by atom if not given an ordering
+
+        Returns: Cohesive Energy
+
+        """
+
+        # determine CN Values
+        (atom_CN, bond_CN) = np.hsplit(self.bond_list, 2)
+        (CN_Vals, CN) = np.unique(bond_CN, return_counts=True)
```

Comment:
Can we convert this to `np.bincount(self.bond_list[:, 0])`? Better yet, will cn ever change when we adjust the ordering? How about you create a CN attribute when you initialize the object: `self.cn = np.bincount(self.bond_list[:, 0])`. That should really boost performance of this method (especially for larger systems).